libtenzir/builtins/operators/cache.cpp

```diff
@@ -0,0 +1,735 @@
+                                           "operators/cache"};
+    auto id = located<std::string>{};
+    auto mode = std::optional<located<std::string>>{};
+    auto capacity = std::optional<located<uint64_t>>{};
+    auto ttl = std::optional<located<duration>>{};
+    auto max_ttl = std::optional<located<duration>>{};
+    parser.add(id, "<id>");
+    parser.add("--mode", mode, "<read|write|readwrite>");
+    parser.add("--capacity", capacity, "<capacity>");
+    parser.add("--ttl", ttl, "<duration>");
+    parser.add("--max-ttl", max_ttl, "<duration>");
+    parser.parse(p);
+    if (mode
+        and (mode->inner != "read" and mode->inner != "write"
+             and mode->inner != "readwrite")) {
+      diagnostic::error("unknown mode `{}`", mode->inner)
+        .note("available modes: read, write, readwrite")
+        .primary(mode->source)
+        .throw_();
+    }
+    if (not capacity) {
+      capacity.emplace(defaults::max_partition_size, location::unknown);
+    }
+    if (not ttl) {
+      ttl.emplace(std::chrono::minutes{1}, location::unknown);
+    } else if (ttl->inner <= duration::zero()) {
+      diagnostic::error("ttl must be a positive duration")
+        .primary(ttl->source)
+        .throw_();
+    }
+    if (not max_ttl) {
+      max_ttl.emplace(duration::zero(), location::unknown);
+    } else if (max_ttl->inner <= duration::zero()) {
+      diagnostic::error("max_ttl must be a positive duration")
+        .primary(max_ttl->source)
+        .throw_();
+    }
+    if (not mode or mode->inner == "readwrite") {
+      auto result = std::make_unique<pipeline>();
+      result->append(std::make_unique<write_cache_operator>(id));
+      result->append(std::make_unique<read_cache_operator>(
+        std::move(id), *capacity, ttl->inner, max_ttl->inner));
+      return result;
+    }
+    if (mode->inner == "write") {
+      return std::make_unique<write_cache_operator>(std::move(id), *capacity,
+                                                    ttl->inner, max_ttl->inner);
+    }
+    TENZIR_ASSERT(mode->inner == "read");
+    return std::make_unique<read_cache_operator>(std::move(id));
+  }
+
+  auto make(invocation inv, session ctx) const
+    -> failure_or<operator_ptr> override {
+    auto id = located<std::string>{};
+    auto mode = std::optional<located<std::string>>{};
+    auto capacity = std::optional<located<uint64_t>>{};
+    auto ttl = std::optional<located<duration>>{};
+    auto max_ttl = std::optional<located<duration>>{};
+    argument_parser2::operator_("cache")
+      .add(id, "<id>")
+      .add("mode", mode)
+      .add("capacity", capacity)
+      .add("ttl", ttl)
+      .add("max_ttl", max_ttl)
+      .parse(inv, ctx)
+      .ignore();
+    auto failed = false;
+    if (mode
+        and (mode->inner != "read" and mode->inner != "write"
+             and mode->inner != "readwrite")) {
+      diagnostic::error("unknown mode `{}`", mode->inner)
+        .note("available modes: read, write, readwrite")
```

Comment:
@raxyte asked today whether we should call the `write` mode `create` instead, as it requires the cache not to exist already (which primarily is the case because of the narrow focus on use cases for the Tenzir Platform). I'm not sure on this, but posting this comment to not lose track of the thought.

Reply:
We had a brief discussion in the team on this earlier today, and as a result I let the name stay `write`, and explicitly explained in the docs that the purpose of this operator is being used in the Tenzir Platform, and that users should generally not use it by themselves.